### DISCUSSION TITLE
NAT traversal

### DISCUSSION THREAD

**Discussion by lostystyg** [pocketnetteam/pocketnet.core#199] [Show and tell]
## Overview
NAT traversal consists of 3 elements:
  - WebRTC endpoint
  - Signaling server
  - Client

### Signaling server
It is a websocket exposed on port 13131 (*probably change and differ for test and main net?*) and **/signaling** endpoint.
It accepts connections and handler it for further redirecting. Each request to signaling server should have a field **"type"**.
Supported types:
  - "protocol" - means this is the webrtc establish protocol and such messages should be simply redirected to requested connection.
```json
{"type": "protocol", "id": "12.213.1.31:12122", "message": {"...protocol message here..."}
```
During redirection, signaling server will replace "id" field from target to sender, so webrtc endpoint will know from who the request comes and how to reply.
The "id" field is used just to specify the target for redirecting and can be changed to anything else.
Answer: *no answer, even if there is no requested id*. The only answer can be from requested webrtc endpoint but no answer from signaling server itself (discuss, can be easily changed).
  - ...
Currently signaling server accepts **every** connection without any validation. 

### WebRTC endpoint

WebRTC endpoint is a compilation of connection to signaling server and performing webrtc establishment to create a datachannel that will be used to perform rpc/websocket requests to the node.
Node tries to connect to a signaling server of every peer it has access to. Basically it tries to have signaling connection with every peer that is shown in "getpeerinfo".
E.x.: **A** is static public node and **B** is private node behind NAT. If "getpeerinfo" rpc call to **A** returns **B**, then **A** is connected with **B** and that means that **B** is also connected with **A**. That means that **B** tried to connect to **A**'s signaling server and if this connection was successful, **B** is now accessible via webrtc. However, there is no way to determine if **B** is really connected to signaling of **A** but at least client can know that this can take place.

Each request to webrtc is accepted from signaling connection.
```json
{"type": "protocol", "id": "12.213.1.31:12312", "message": {"...protocol message here..."}
```
see 'Signaling server' for details.

Format of "message":
Each "message" object is identified by a "type" field that can be one of the following:
- "offer": an initiate request from client that starts webrtc protocol
```json
{ "type": "offer", "sdp": "session description text"}
```
"sdp" - a session description that should be created by `peerConnection.createOffer()` call on client.
Answer - webrtc answer sdp:
```json
{ "type": "answer", "sdp": "...session description string here..." }
```
- "candidate": a part of ICE-trickle webrtc candidate exchange
```json
{ "type": "candidate", "candidate" : "...candidate data string here...", "mid": "...mid string here..." } 
```
"candidate" - candidate description. If obtained from `onicecandidate` event from js webrtc it should equal to `event.candidate.candidate`.
"mid" - if obtained from `onicecandidate` event from js webrtc it should equal to `event.candidate.sdpMid`.

After peer connection established, webrtc endpoint will accept datachannels that should be created from client by `peerConnection.createDataChannel('label')` call.
The label will identify the type of requests (rpc or websocket). See 'Communication protocol over data channel' for details.

### Client
On proxy level client "pings" every node to identify if it is connectable or not.
At this level, if node **B** is not connectable via common connection, client should also try to connect to it via webrtc and signaling and mark it somehow if successful. As a signaling client should use the node **A** that returns **B** in `getpeerinfo`. Client should also be aware that there will be more than one node that will return **B** in `getpeerinfo`.
Client should decide which node to connect based on the "pings" described above and on some statistic that can be obtained from node via `getnodeinfo` (*these stats are not yet implemented*).
Client is always an initiator of webrtc protocol (always send an offer) and format of messages is the same as described for 'Webrtc endpoint'.
Client should handle webrtc connection for entire session (not creating new one for each request as it is done for http rpc).

## Notes
1. webrtc is using [SCTP protocol for sending arbitrary data](https://webrtc.org/getting-started/data-channels) between peers. It has kept connection alive until socket destroyed but have some long delay (around 1-2 mins) until the other side is notified about connection was dropped.
2. Each node tries to connect to each node's signaling server and that means that even public static node can be accessible via webrtc.

### Listing available nodes
```json
{
  "type": "listnodes"
}
```
Answer: list of nodes identifiers.

This request will show only nodes signaling connections (excluding clients) and currently identifier is equal to the `addrLocal` of the that peer. **NOTE**: this can be changed!!!

#### RPC
Currently it is working like this:
  - Http rpc requests differ on URI part of the request.
    - e.x. call to `1.1.1.1:38081/` and `1.1.1.1:38081/post/` will allow to access different rpc calls.
  - Http rpc request are all POST and have a body.

To not change the logic of rpc processing and request forming, for data channel protocol i suggest just encapsulate http request's body and add additional parameter `path` that will represent the URI part of http request.
Also because communication via datachannel is fully async, client must generate unique ID for each request and the node will respond with same ID.
So if http request is:
```code
POST /post/
Host: ...

{ "method": "getnodeinfo", "params": [...]}
```
and response:
```
{ "key": "value", ...}
```

The data channel request will be:
```code
{
  "path": "/post/",
  "id": "some unique request id as string",
  "requestData": {"method": "getnodeinfo", "params": [...]}
}
``` 
and response:
```
{ 
  "id": "some unique request id as string",
  "data": { "key": "value", ... }
}
```
Http result codes are ignored for datachannel
#### Websocket
Not yet implemented but should be the same as currently.

**Comment by andyoknen** [comment]
2.2 (3) - Nodes must provide job statistics in getnodeinfo and getpeerinfo. Ideally, electron clients should use NAT nodes to leave public web clients

**Comment by andyoknen** [comment]
Notes:
1 - The RPC holds the connection for a while until the request is processed by the queue. After the response, the connection is reset. Because p2p assumes a permanent connection of the client to the node, we can put all incoming requests in the queue, leave a note that the request came via WS and send the response to WS, instead of responding via HttpResponse

**Comment by lostystyg** [comment]
Some notes about possible bugs from official samples repo: https://github.com/webrtc/samples/blob/0a788d6c3c2bee443ea6a2c1675fa15a3f8d9fa3/src/content/datachannel/datatransfer/js/main.js#L200

**Reply by andyoknen** [reply]
A strange workaround is to close the socket)
